pilot/cmd/pilot-agent/status/server.go

```diff
@@ -508,29 +527,33 @@ func (s *Server) handleStats(w http.ResponseWriter, r *http.Request) {
 		format = expfmt.FmtText
 	}
 
-	if agent, err = scrapeAgentMetrics(); err != nil {
-		log.Errorf("failed scraping agent metrics: %v", err)
-		metrics.AgentScrapeErrors.Increment()
-	}
-
 	w.Header().Set("Content-Type", string(format))
 
 	// Write out the metrics
-	if _, err := w.Write(agent); err != nil {
-		log.Errorf("failed to write agent metrics: %v", err)
+	if err = scrapeAndWriteAgentMetrics(io.Writer(w)); err != nil {
+		log.Errorf("failed scraping and writing agent metrics: %v", err)
 		metrics.AgentScrapeErrors.Increment()
 	}
+
+	buf := bufPool.Get().(*[]byte)
+	defer bufPool.Put(buf)
+
 	if envoy != nil {
-		if _, err := w.Write(envoy); err != nil {
-			log.Errorf("failed to write envoy metrics: %v", err)
+		_, err = io.CopyBuffer(w, envoy, *buf)
```

Comment:
I believe this is not needed - we can do without buffer entirely.

`w` is a `*http.response` which implements `io.ReaderFrom`. In CopyBuffer, there is an optimatization for ReaderFrom which ignores the `buf` entirely (because it uses its own buffer pool under the hood).

As a result, I think this probably does nothing except in unit tests/benchmarks (where w does *not* implement ReaderFrom).

I think in real world we can drop the bufpool and simplify this a bit by using `io.Copy`

Reply:
> w is a *http.response which implements io.ReaderFrom.

Thanks. I just found this out.